server/dial_server.c

```diff
@@ -491,69 +491,100 @@ static void handle_dial_data(struct mg_connection *conn,
     ds_unlock(ds);
 }
 
-static int ends_with(const char *str, const char *suffix) {
-    if (!str || !suffix)
-        return 0;
-    size_t lenstr = strlen(str);
-    size_t lensuffix = strlen(suffix);
-    if (lensuffix > lenstr)
+
+static int host_matches(const char *origin_host, const char *host) {
+    if (!origin_host || !host)
         return 0;
-    return strncmp(str + lenstr - lensuffix, suffix, lensuffix) == 0;
+        
+    // If the host contains a port number (indicated by a colon) require an
+    // exact match.
+    const char * origin_colon = strrchr(origin_host, ':');
+    const char * host_colon = strrchr(host, ':');
+    if (host_colon != NULL) {
+        // If the host port number is 443, then accept the origin host if it
+        // does not have any port number under the assumption we already
+        // verified https://
+        if (strlen(host_colon) == 4 &&
+	    strncmp(host_colon, ":443", 4) == 0 &&
+	    origin_colon == NULL)
+	{
+	    // Strip off the host port number and require an exact match.
+  	    size_t host_len = host_colon - host;
+	    if (host_len == 0)
+   	        return 0;
+	    return strncmp(origin_host, host, host_len) == 0;
+        }
+        
+        // Other port numbers must match exactly.
+    	if (strlen(origin_host) != strlen(host))
+    	    return 0;
+    	return strncmp(origin_host, host, strlen(host)) == 0;
+    }
+    
+    // Otherwise strip off any port number in the origin host, and require an
+    // exact match.
```

Comment:
this comment is kind of weird, because you're not requiring an exact match (I get you're trying to say require an exact match on the origin without the port number, but it's phrased weird).